astropy/nddata/nddata.py

```diff
@@ -299,5 +299,13 @@ def uncertainty(self, value):
             # If it is a subclass of NDUncertainty we must set the
             # parent_nddata attribute. (#4152)
             if isinstance(value, NDUncertainty):
+                # In case the uncertainty already has a parent create a new
```

Comment:
Yes, without it I get a lot of these errors in the tests:

```
______ test_arithmetics_stddevuncertainty_with_units[uncert15-uncert25] _______

uncert1 = StdDevUncertainty([ 1.,  2.,  3.])
uncert2 = StdDevUncertainty([2, 3, 4])

    @pytest.mark.parametrize(('uncert1', 'uncert2'), [
        (np.array([1, 2, 3]) * u.m, None),
        (np.array([1, 2, 3]) * u.cm, None),
        (None, np.array([1, 2, 3]) * u.m),
        (None, np.array([1, 2, 3]) * u.cm),
        (np.array([1, 2, 3]), np.array([2, 3, 4])),
        (np.array([1, 2, 3]) * u.m, np.array([2, 3, 4])),
        (np.array([1, 2, 3]), np.array([2, 3, 4])) * u.m,
        (np.array([1, 2, 3]) * u.m, np.array([2, 3, 4])) * u.m,
        (np.array([1, 2, 3]) * u.cm, np.array([2, 3, 4])),
        (np.array([1, 2, 3]), np.array([2, 3, 4])) * u.cm,
        (np.array([1, 2, 3]) * u.cm, np.array([2, 3, 4])) * u.cm,
        (np.array([1, 2, 3]) * u.km, np.array([2, 3, 4])) * u.cm,
        ])
    def test_arithmetics_stddevuncertainty_with_units(uncert1, uncert2):
        # Data has same units
        data1 = np.array([1, 2, 3]) * u.m
        data2 = np.array([-4, 7, 0]) * u.m
        if uncert1 is not None:
            uncert1 = StdDevUncertainty(uncert1)
            if isinstance(uncert1, Quantity):
                uncert1_ref = uncert1.to(data1.unit).value
            else:
                uncert1_ref = uncert1
            uncert_ref1 = StdDevUncertainty(uncert1_ref, copy=True)
        else:
            uncert1 = None
            uncert_ref1 = None

        if uncert2 is not None:
            uncert2 = StdDevUncertainty(uncert2)
            if isinstance(uncert2, Quantity):
                uncert2_ref = uncert2.to(data2.unit).value
            else:
                uncert2_ref = uncert2
            uncert_ref2 = StdDevUncertainty(uncert2_ref, copy=True)
        else:
            uncert2 = None
            uncert_ref2 = None

        nd1 = NDDataArithmetic(data1, uncertainty=uncert1)
        nd2 = NDDataArithmetic(data2, uncertainty=uncert2)

        nd1_ref = NDDataArithmetic(data1, uncertainty=uncert_ref1)
        nd2_ref = NDDataArithmetic(data2, uncertainty=uncert_ref2)

        # Let's start the tests
        # Addition
        nd3 = nd1.add(nd2)
        nd3_ref = nd1_ref.add(nd2_ref)
        assert nd3.unit == nd3_ref.unit
        assert nd3.uncertainty.unit == nd3_ref.uncertainty.unit
        assert_array_equal(nd3.uncertainty.array, nd3.uncertainty.array)

>       nd3 = nd2.add(nd1)

astropy\nddata\mixins\tests\test_ndarithmetic.py:585:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
astropy\nddata\mixins\ndarithmetic.py:528: in add
    **kwargs)
astropy\nddata\mixins\ndarithmetic.py:612: in _prepare_then_do_arithmetic
    result, init_kwds = operand._arithmetic(operation, operand2, **kwargs)
astropy\nddata\mixins\ndarithmetic.py:272: in _arithmetic
    **kwds2['uncertainty'])
astropy\nddata\mixins\ndarithmetic.py:407: in _arithmetic_uncertainty
    correlation)
astropy\nddata\nduncertainty.py:269: in propagate
    correlation)
astropy\nddata\nduncertainty.py:493: in _propagate_add
    if self.unit != other_uncert.unit:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = StdDevUncertainty([2, 3, 4])

    @property
    def unit(self):
        """`~astropy.units.Unit` : The unit of the uncertainty, if any.

            If the unit is not set the unit of the parent will be returned.

            .. warning::

              Setting or overwriting the unit manually will not check if the new
              unit is compatible or convertible to the old unit. Neither will this
              scale or otherwise affect the saved uncertainty. Appropriate
              conversion of these values must be done manually.
            """
        if self._unit is None:
            if (self._parent_nddata is None or
>                   self.parent_nddata.unit is None):
E                   AttributeError: 'NoneType' object has no attribute 'unit'

astropy\nddata\nduncertainty.py:152: AttributeError
```
That was a whole lot of investigation to find out why the following `if` was required. I can give you more details if you want.

Note that it also resolves some problems like this:
```
ndd1 = NDData(1, uncertainty=1)
ndd2 = NDData(2, uncertainty=ndd1.uncertainty)
ndd1.uncertainty.parent_nddata.data # = array(2) 
# so the uncertainty of ndd1 now thinks it's parent is ndd2
```
with the fix it returns the expected `array(1)`.